fitbenchmarking/core/results_output.py

```diff
@@ -199,6 +201,183 @@ def preprocess_data(results: "list[FittingResult]"):
     return best_results, sorted_results
 
 
+def _handle_fallback_tags(results,
+                          all_result_tags,
+                          fallback_columns,
+                          col_order):
+    """
+    The function that relabels the fallback column tags that
+    appear when there is jacobian and hessian fallbacks for a
+    minimizer.
+
+    :param results: The list of results of benchmarking
+    :type results: list[FittingResult]
+    :param all_result_tags: A list of tags that can be used
+                            to sort the results
+    :type: list[dict[str, str]]
+    :param fallback_columns: The col tag of the fallback columns
+    :type fallback_columns: list[str]
+    :param col_order: The sort order of the col tags
+    :type col_order: list[str]
+
+    :return: all results and the results tags
+    :rtype: list[FittingResult], list[dict[str, str]]
+    """
+    column_rename = "best_avaliable"
+
+    rows_count = len({row['row'] for row in all_result_tags})
+    col_tags = [tag['col'] for tag in all_result_tags]
+    all_count = {key: col_tags.count(key)
+                 for key in set(col_tags)}
+
+    for ix, tag in enumerate(all_result_tags):
+        # If tag is in fallback columns list
+        if tag['col'] in fallback_columns:
+
+            result_ix = tag["result_ix"]
+
+            # Unpack the tag
+            unpacked_column_tag = dict(zip(col_order, tag["col"].split(":")))
+
+            # Find matches for jacobian and hessian tags
+            matches_summary = {}
+            for check_tag in ['jacobian', 'hessian']:
+                match_str = ':'. join(["[^:]*" if key == check_tag else
+                                      unpacked_column_tag[key]
+                                      for key in col_order])
+                matches = _find_matching_tags(match_str, fallback_columns)
+                matches_summary[check_tag] = matches
+
+            rename_jac_and_hes = len(matches_summary['jacobian']) == 2 \
+                and len(matches_summary['hessian']) == 2
+
+            if rename_jac_and_hes:
+                # If both jacobian and hessian tags can be renamed
+                matches_summary['jacobian'].remove(tag['col'])
+                col_count = all_count[tag['col']]
+                jac_count = all_count[''.join(matches_summary['jacobian'])]
+
+                # If jacobian renaming leads to row count matching
+                if jac_count + col_count == rows_count:
```

Comment:
One worry with this is that we could have an unfortunate even number:

|problem | j1-h1 | j1-h2 | j2-h1 | j2-h2 | description|
|-|-|-|-|-|-|
|p1 | ❌ | ✔️ | ❌ | ✔️ | hess fallback |
|p2 | ❌ | ✔️ | ❌ | ✔️ | hess fallback |
|p3 | ✔️ | ❌ | ✔️ | ❌ | analytic hess |
|p4 | ✔️ | ❌ | ✔️ | ❌ | analytic hess |

I think this logic would try to combine the j1 columns?

Reply:
Yes, It will try to combine the j1 and j2 tags. In this case, we would have two columns tags

- `best_avaliable:h1`
- `best_avaliable:h2`. 

The following part of the code would decide whether it should update the jacobian or the hessians. But the preference is given to the jacobian update.

However, this was an arbitrary choice. We can prioritize hessian renaming so that we end up with

- `j1:best_avaliable`
-  `j2:best_avaliable`

```
# If both jacobian and hessian tags can be renamed
matches_summary['jacobian'].remove(tag['col'])
col_count = all_count[tag['col']]
jac_count = all_count[''.join(matches_summary['jacobian'])]

# If jacobian renaming leads to row count matching
if jac_count + col_count == rows_count:
      # Update jacobian tag
      unpacked_column_tag['jacobian'] = \
            results[result_ix].jacobian_tag = column_rename
else:
       # Update hessian tag
       unpacked_column_tag['hessian'] = \
             results[result_ix].hessian_tag = column_rename
```

In my opinion both are okay. I will try to add a test_case for this as well. However, It is similar to edgecase2.json.  I think it should work even if it is an even number.

Reply:
I think the problem is we'd need to know where the missing results are (the ❌'s).
So `if (jac_count + col_count == rows_count) and ((results[result_ix] is not None) != (results[fallback_col_ix] is not None))` then do the jac column?

Reply:
I think you are right. The logic should automatically identify the right columns to match. In this case `best_avaliable:h1` and `best_avaliable:h2.` is the wrong answer. I have added this edgecase to the tests.